mne/beamformer/_ap.py

```diff
@@ -0,0 +1,712 @@
+            sub_g = gain[:, s_ap_temp]
+            act_spc = _active_subspace(sub_g)
+            perpend_spc = np.eye(act_spc.shape[0]) - act_spc
+            for dip in range(attr_dict['ndipoles']):
+                l_p = np.expand_dims(gain[:, dip], axis=1)
+                ap_val2[dip] = multi_dot(
+                    [l_p.T, perpend_spc, data_cov, perpend_spc, l_p]
+                ) / ((multi_dot([l_p.T, perpend_spc, l_p]))[0, 0])
+            s2_idx = np.argmax(ap_val2)
+            s_ap_2[src] = s2_idx
+        logger.info('current s_ap_2 = {}'.format(s_ap_2))
+        if (itr > 0) & (s_ap_2_prev == s_ap_2):
+            # No improvement vs. previous iteration
+            logger.info('Done (optimally)')
+            break
+        if itr == attr_dict['max_iter']:
+            logger.info('Done (max iteration)')
+    return s_ap_2
+
+
+def _calculate_fixed_alternating_projections(data_arr, gain,
+                                             nsources,
+                                             max_iter):
+    """Calculate fixed-orientation alternating projection.
+
+    Parameters
+    ----------
+    data_arr : array, shape (nchannels, times)
+        Filtered evoked data array.
+    gain : array, shape (nchannels, ndipoles)
+        Gain matrix.
+    nsources : int
+        The number of dipoles to estimate.
+    max_iter : int
+        Maximal iteration number of AP.
+
+    Returns
+    -------
+    s_ap_2 : list of int
+        List of dipole indices.
+
+    """
+    s_ap = []
+    ndipoles = gain.shape[1]
+    attr_dict = {
+        'ndipoles': ndipoles,
+        'nsources': nsources,
+        'max_iter': max_iter
+    }
+    logger.info('calculating fixed-orientation alternating projection')
+    data_cov = _produce_data_cov(data_arr, attr_dict)
+
+    # ######################################
+    # 1st Phase
+    # (a) Initialization: search the 1st source location over the entire
+    # dipoles topographies space
+    # ######################################
+
+    logger.info(' 1st phase : ')
+    s_ap = _fixed_phase1a(attr_dict, data_cov, gain)
+
+    # ######################################
+    # (b) Now, add one source at a time
+    # ######################################
+
+    s_ap = _fixed_phase1b(gain, s_ap, data_cov, attr_dict)
+
+    # #####################################
+    # 2nd phase
+    # #####################################
+
+    logger.info(' 2nd phase : ')
+    s_ap_2 = copy(s_ap)
+    s_ap_2 = _fixed_phase2(attr_dict, s_ap_2, gain, data_cov)
+
+    return s_ap_2
+
+
+def _solve_active_gain_eig(ind, gain, data_cov, eig, perpend_spc):
+    """Eigen values and vector of the projection."""
+    gain_idx = list(range(ind * 3, ind * 3 + 3))
+    l_p = gain[:, gain_idx]
+    eig_a = multi_dot(
+        [l_p.T, perpend_spc, data_cov, perpend_spc, l_p])
+    eig_b = multi_dot(
+        [l_p.T, perpend_spc, perpend_spc, l_p])
+    eig_b = eig_b + 1e-3 * eig_b.trace() * np.eye(3)
```

Comment:
```suggestion
    # Add to small regularisation on the diagonal
    eig_b[::3 + 1] += 1e-3 * eig_b.trace()
```
Is this parameter mentioned in the paper? Should it deserve an explanation here?

Reply:
Quoting from a correspondance with Dr. Amir Adler on your inquary:
"The generalized eigenvalue problem with the matrix pencil [F,G] is equivalent to the following standard eigenvalue problem:  G^(-1)*F*v= lambda * v
Therefore, in order to ensure non-singularity of G (as defined in eq. (23) in the manuscript), I added the diagonal loading"
where G is eig_b in code.